### ISSUE TITLE
sam: Consider more conventional format versioning

### ISSUE THREAD

**Issue by zaeleus** [samtools/hts-specs#725]
This is in regard to _Sequence Alignment/Map Format Specification_ (2022-08-02) § "Appendix B: SAM Version History".

Specification changes are currently made to the latest version. This is both confusing and unconventional, as versioned formats should be considered immutable.

I propose 1.6 be finalized and future changes made to a draft of the next version of SAM.

**Comment by jmarshall** [comment]
The **format** itself **is** versioned in a conventional manner: generally when syntax is introduced that previous implementations would fail to parse, the version number is bumped to indicate that. This is indicated by the bold items in Appendix B. For example, when `B` arrays were added as a new form of optional field, the `@HD VN` version number was bumped to 1.4 accordingly.

The SAM format is designed to be an **extensible** format. Thus it defines syntax for various kinds of optional SAM record fields, and invites tools and research groups to define meanings for particularly named tagged fields. Similarly it defines (simpler untyped) syntax for SAM header fields, and invites users to define meanings for particularly named fields. There are also particular tagged fields and header fields that are predefined by the SAM format authors and maintainers because they (and the shared vocabulary they define) are of general use; for convenience, these are listed in the same specification documents.

There is no issue with a particular tagged or header field defined later appearing in a SAM/BAM/etc file that declares an earlier `@HD VN` version (provided it is not a `B` array in a VN 1.3 file or a similar scenario, of course), as it is really just another field. For example, `@SQ AH` and `MI:Z` were defined in 2017 while the SAM version number was 1.5, but it is perfectly appropriate for them to appear in a file that declares `VN:1.3` — it would not benefit users if tools rejected such a scenario.

Similar comments apply to other **semantic** information such as the list of (valid) well-known values for such fields, as mentioned in https://github.com/samtools/hts-specs/issues/718#issuecomment-1528780718.

This means that the specification document contains a mixture of categories of information: there is the format proper that determines syntax and what implementations need to parse, and semantic definitions of particular tags perhaps accompanied by lists of valid values and their definitions. (The definitions of optional record fields used to be a section within _SAMv1.pdf_ too BTW. _SAMtags.pdf_ was moved to a separate document for two main reasons: so that it could be more reasonably referred to from the CRAM specification; to somewhat decouple it and its maintenance from the SAM document and its versioning. These header field definitions could be moved to _SAMtags.pdf_ too for similar reasons, but IMHO that would be overkill — I have a backburner PR to reformat the multipage header table as not-a-table, which would mean we had more space to expand on what “valid values” really means and clarify this.)

> versioned formats should be considered immutable

What I think you are suggesting implies that absolutely everything defined in the specification documents would no longer be considered extensible, in the sense that fields and valid values defined in a particular edition of the specification would not be allowed in SAM files that predated that edition (or that declared an `@HD VN` value prior to that edition). Also the specification would need to define much more fine-grained version numbers for use in `VN`.

The question is who this would be an improvement for. There are at least three categories of people with an interest in these specifications:

* Users of SAM/BAM/etc files and of tools that deal with these files
* Implementors of tools that deal with these formats
* The specification authors

It is true that tools that validate things carefully may wish to say things like “I support PL:ULTIMA because I support edition XYZ of the SAM specification”. But SAM philosophy has always been that such statements are not particularly useful for the users working with SAM files, and that where it means that otherwise compatible new fields can't be used in old files it's actively counterproductive for users.

TL;DR There is a difference between versioning of the format and versioning of the specification document. This has been discussed ad nauseam before, for example on #194 and #454.

In practice, specification documents are not as immutable as you might think. Typo fixes and formatting improvements do not change the format described by a specification. Clarification of unclear text should not and generally does not change the format described by a specification. Thus not all changes in a specification document should be reflected in the version number declared by a file in that format. (These are examples of when specification documents are not usefully considered immutable.)

Hence the SAM specification separates the SAM version number (like the current 1.6) from the specification's own versioning. The specification document itself is lightly versioned: its title page mentions the Git commit hash it was built from and the date of that commit. This light versioning has IMHO proved sufficient for readers' needs.

That is the status quo.

If the status quo were to be revisited, those advocating for a change ought to familiarise themselves with the previous discussions and consider the following:

* Do you accept the idea of SAM as an **extensible** format and the implications of that described above?
* Do you accept the distinction between syntax and semantics described here and on e.g. #718? (Perhaps you may have suggestions on making the distinction in the spec clearer.)
* What, specifically, would be your proposal, and what actual benefits would it have for users, for implementors, and for the specification editors?

**Comment by jmarshall** [comment]
>> Do you accept the distinction between syntax and semantics described here[…]
>
> No, I consider this all part of the grammar.

Well, that is a view that is not held by others. It is IMHO an implementor-centric view, and not one that is especially useful for users of SAM or for most implementors of bioinformatics tools.

If it is not clear that things like the lists of valid well-known values for `PL` are semantics information and of a different category from the grammar proper, then that is something to be clarified in the spec. I outlined how I think we should do this in parentheses in my previous comment.

Certainly there is a tension here for implementations that want to validate PL values: it is useful that there exist implementations that produce diagnostics for unknown values, because that is how we get consistency in the values written by users; but there is a deployment issue for such implementations when new platforms arrive and are added to the spec. The spec just lists the canonical values (imperfectly) and leaves decisions about whether to validate this field value up to the implementations. Users would not be helped by having to set `@HD VN` to some inflated value if they used PL:ELEMENT, and having their SAM files rejected by (some) tools if they did not. 

> Your first example is a counterexample to your argument:
>
>> For example, when B arrays were added as a new form of optional field, the `@HD VN` version number was bumped to 1.4 accordingly.
>
> TAG:TYPE:VALUE is, as you call, the syntax. TYPE is "a single case-sensitive letter". Therefore, TAG:B:VALUE was already accepted. Given that TYPE is a token and not an invariant, VALUE can be arbitrary. Why was this able to influence the version bump?

With our spec maintainers hats on, we need to think about what SAM as a Platonic ideal is and means, and what the spec text means to say and ought to say, not just what the imperfect text as it currently stands actually says.

Here `TYPE` is “a single case-sensitive letter **from the following table**”, whether the text says that explicitly or not. The defined TYPE letters are a different category of thing to the PL values, because they define how to interpret VALUE and what syntax is expected when parsing VALUE,[^1] and they define how the field is represented in BAM. Certainly an implementation would have no hope of parsing a **BAM** record that contained a B array prior to the addition of B to the specification or prior to the addition of specific support for B arrays as defined in the spec to the implementation. Hence the defined TYPE letters and the implications for representation of VALUE in SAM and BAM are part of the syntax, in a way that the list of PL values is not.

In the light of those considerations, it is obvious that TAG:B:VALUE could not already have been accepted. Hence the version bump.

[^1]: I certainly agree that regexes by themselves are an insufficient way of describing what things specified via regex are supposed to mean.

**Comment by jmarshall** [comment]
> Changing the behavior of the format within the same version is not conventional and confusing for all parties (except the spec maintainers?).

The spec maintainers would claim that they have not substantially changed the behaviour of the format within the same version.

Your views on PL are clear. You do have a point and the description of this field should be expanded and improved in some approximation of the ways talked about.

Are there other specific problems you have in mind? (Or is this a storm in a PL-sized teacup…?)

**Comment by jkbonfield** [comment]
> `TAG:TYPE:VALUE` is, as you call, the syntax. `TYPE` is "a single case-sensitive letter". Therefore, `TAG:B:VALUE` was already accepted. Given that `TYPE` is a token and not an invariant, `VALUE` can be arbitrary. Why was this able to influence the version bump?

SAM and BAM are two different specs, with different version numbers, but there is some rather poorly defined tie between them.  BAM has its own binary version number, but that is more for the higher encapsulation and encoding of fixed fields.  There is also the SAM version number within the text header within BAM.

For B, arguably both BAM and SAM could have changed version, but given BAM encapsulation elsewhere doesn't change and the SAM header version can be specified within BAM, I guess we decided to stick with SAM only.

An example with B:

In SAM, it's just an array.  Basically it's just a new type as you say.  The syntax doesn't change, and we can still skip from SAM tag to SAM tag trivially by looking for the tabs and recognising TAG:TYPE:VALUE.  Even if we get a TYPE we don't know, our parser could be written to be liberal and just skip over the tag, or store it in pure text and regurgitate it back out as-is without any knowledge of the new type code.

In BAM this is fundamentally impossible.  BAM tags are all squished together without a separator like TAB.  We *must* know what the new TYPE means and how to parse it, otherwise our entire file is unparsable.  To that degree, the B tag is a hard stop.  It's not just a case of interpretation of data and recognising new key-value names, but of the fundamental data layout.

I agree it's nuanced, and maybe not desireable, to have both BAM and SAM intertwined in this way, but it's where we are and we cannot change history.  So do remember this when thinking in pure abstract terms of syntax and semantics - we're not just in SAM world.
